### DISCUSSION TITLE
re-render tree Data with Uncontrolled Environment Tree

### DISCUSSION THREAD

**Discussion by mansi-softvan** [lukasbach/react-complex-tree#18] [Show and tell]
Hey, I'm facing the issue of re-render the Uncontrolled environment tree. while adding new data into the list then the list will be updated but does not reflect in the tree.

I have attached the video of this issue. kindly check it and tell us Is there an option for re-rendering the tree with an uncontrolled environment? 

https://user-images.githubusercontent.com/66248382/148977003-93edfaab-debf-47f8-a5e0-677c3ece0930.mp4

Thanks in Advance!

**Comment by doMynation** [comment]
I believe you have to use a [controlled environment](https://rct.lukasbach.com/docs/guides/controlled-environment) to have your tree react to changes applied to your data. Either that, or implement a custom data provider that listens to changes in your data.

**Comment by lukasbach** [comment]
Yes, as @doMynation said, either you need to provide the data in a controlled environment, or what is likely easier in your case, use a custom dataprovider that implements the `onDidChangeTreeData` method, then use that to manage change listeners that you trigger to notify the tree environment on changes in the data structure.

**Comment by PooLP** [comment]
Do you have an example of how to implement a custom Dataprovider / onDidChangeTreeData  for UncontrolledTreeEnvironment?
With my test, I get all items, but Drag&Drop functionality do not work anymore.

**Reply by lukasbach** [reply]
Does the implementation of the default dataprovider help? https://github.com/lukasbach/react-complex-tree/blob/main/packages/core/src/uncontrolledEnvironment/StaticTreeDataProvider.ts

**Reply by PooLP** [reply]
Thanks. 
Note for me. look the source code 😄

**Comment by perkins2099** [comment]
Still struggling getting the tree to recognized refreshed tree data.  I made a new function in a copy of the StaticTreeDataProvider 

```
  public refreshData(dataIn){
    this.data.items = dataIn
  }
```

This can change the items data / name and it renders in the ui but the above doesnt.  
```
    Object.keys(dataIn).forEach(key => {      
      this.data.items[key].data.name = "Changed"
    })
```

But it looks like the data is already being updated but the UI isn't rendering the updates.

Anyone get this working?

Thanks
